samples/NativeCatalog/android/app/src/main/java/com/pspdfkit/nativecatalog/CustomPdfViewManager.java

```diff
@@ -0,0 +1,309 @@
+                .signatureSavingStrategy(SignatureSavingStrategy.NEVER_SAVE)
+                .autosaveEnabled(false)
+                .build();
+            pdfView.setConfiguration(config);
+
+            // The pdf view needs a fragment tag to function, make sure this is unique inside your app.
+            pdfView.setFragmentTag("pdfView");
+
+            // Finally return it to react-native.
+            return pdfView;
+        }
+
+        throw new IllegalStateException("CustomPdfView can only be used in FragmentActivity subclasses.");
+    }
+
+    // This will allow us to set the document via react-native.
+    @ReactProp(name = "document")
+    public void setDocument(PdfView view, @NonNull String document) {
+        view.setDocument(document);
+    }
+
+    @Nullable
+    @Override
+    public Map getExportedCustomDirectEventTypeConstants() {
+        // These are required for the PdfView to work.
+        // You can see them in ReactPdfViewManager.java
+        Map map = MapBuilder.of(PdfViewStateChangedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onStateChanged"),
+            PdfViewDocumentSavedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaved"),
+            PdfViewAnnotationTappedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationTapped"),
+            PdfViewAnnotationChangedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationsChanged"),
+            PdfViewDataReturnedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDataReturned"),
+            PdfViewDocumentSaveFailedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaveFailed"),
+            PdfViewDocumentLoadFailedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentLoadFailed")
+        );
+        // This is the event that is sent after digitally signing.
+        map.put(DocumentDigitallySignedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentDigitallySigned"));
+        // This is the event that is sent after watermarking.
+        map.put(DocumentWatermarkedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentWatermarked"));
+        return map;
+    }
+
+    @Override
+    public void receiveCommand(@NonNull PdfView root, String commandId, @Nullable ReadableArray args) {
+        super.receiveCommand(root, commandId, args);
+        switch (commandId) {
+            case "startSigning":
+                // The user clicked the signing button in react-native, start the native signing flow.
+                performInkSigning(root);
+                break;
+            case "createWatermark":
+                // The user clicked the create watermark button, start our watermarking process.
+                performWatermarking(root);
+                break;
+        }
+    }
+
+    private void performInkSigning(@NonNull PdfView pdfView) {
+        pdfView.getFragment()
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(pdfFragment -> {
```

Comment:
Why is fragment retrieval asynchronous? Is this encapsulating some kind of lifecycle management?

Reply:
Yes, since the `PdfFragment` will only be created by the `PdfView` once all props have been set we need to handle this asynchronously.